### ISSUE TITLE
Add New Show not saving images to cache

### ISSUE THREAD

**Issue by Hiltronix** [SiCKRAGETV/sickrage-issues#2547]
For the past week, when I add a new show to SR, the image files; posters, banners, etc, are saved to the video destination directory, but are not being saved to the cache/images directory.  So poster images are not displayed in the web UI for newly added shows as they used to.

Log:
https://gist.github.com/Hiltronix/47b7a883e824a4dbbd61#file-add-new-show-not-saved-in-cache

Windows 7 64
Python 2.7 64
SR Version: BRANCH: (master) / COMMIT: (5770d33ef96e619af764b452ea35ad33e5755add)

**Comment by miigotu** [comment]
@Hiltronix this is something wrong with your image cache.

`2015-08-23 12:05:54 ERROR    TORNADO :: API :: error 9 : Bad file descriptor`

**Comment by Hiltronix** [comment]
Okay.  This isn't an issue I've had to solve before.  Do you have any suggestions on how I can fix an image cache problem?

**Comment by WebSpider** [comment]
Usually:

1. Stop SR
2. Rename/delete cache/images
3. Restart SR
4. Force a full update

**Comment by cookerrr** [comment]
I am also having this same issue.  Of the 80 some shows about 10 show the banners/posters in the sickrage interface.  I've erased the cache and forced a full update but it doesn't help.  the images are downloaded by sickrage and put in the show directories, just not coped to the cache.

**Comment by cookerrr** [comment]
Here is a log from a forced update for the show amerca's got talent.  All the artwork has been put in the folder by sickrage, but sickrage didn't copy it to its cache folder nor display it in the sickrage interface.

https://gist.github.com/cookerrr/eabf1989d2f7a6e2bbde

Any ideas?  It seems like the last entry is saying it wants to use the banner/poster from the metadata it created but then just stops doing anything.  Am I doing something wrong?

**Comment by cookerrr** [comment]
just to add to my other posts.  when I take a poster from a show directory and rename it to <show id>.poster.jpg and move it to the cache/images folder sickbeard will start displaying that show banner in the interface.  Is the poster view and artwork something new that was just added that's a work in progress or should this be working correctly?  I still can't figure out why a few of my shows imported the poster and banner to the cache/images directory but the rest haven't.

**Comment by cookerrr** [comment]
Let me preface this by saying I know 0 python, but i was poking around in images_cache.py in the python directory and found the section at the bottom where sickrage would look for current images in the metadata it makes  Below is the portion I'm speaking of from the file, which is the last 50 or so lines.  If i delete the portion that looks for images in the metadata, it'll go and grab the images from thetvdb successfully and put them in the cache just fine.  Any idea why the code for using the existing images doesn't work?  I'm on windows 7, does that matter for os.path.isfile?  From the log I posted above it seems to get to checking if we can use show image from and then the script stops, because the log doesn't show any other info.  For now my issue is fixed because it just pulls missing images from thetvdb, but not sure why it's not working correctly in the first place.  Thanks for any help!

       if need_images[self.POSTER] or need_images[self.BANNER] or need_images[self.FANART]:
            try:

                for cur_provider in sickbeard.metadata_provider_dict.values():
                    logger.log(u"Checking if we can use the show image from the " + cur_provider.name + " metadata",
                               logger.DEBUG)
                    if ek.ek(os.path.isfile, cur_provider.get_poster_path(show_obj)):
                        cur_file_name = os.path.abspath(cur_provider.get_poster_path(show_obj))
                        cur_file_type = self.which_type(cur_file_name)
                        if cur_file_type == None:
                            logger.log(u"Unable to retrieve image type, not using the image from " + str(cur_file_name),
                                       logger.WARNING)
                            continue

                        logger.log(u"Checking if image " + cur_file_name + " (type " + str(
                            cur_file_type) + " needs metadata: " + str(need_images[cur_file_type]), logger.DEBUG)

                        if cur_file_type in need_images and need_images[cur_file_type]:
                            logger.log(
                                u"Found an image in the show dir that doesn't exist in the cache, caching it: " + cur_file_name + ", type " + str(
                                    cur_file_type), logger.DEBUG)
                            self._cache_image_from_file(cur_file_name, cur_file_type, show_obj.indexerid)
                            need_images[cur_file_type] = False
            except exceptions.ShowDirNotFoundException:
                logger.log(u"Unable to search for images in show dir because it doesn't exist", logger.WARNING)

        # download from indexer for missing ones
        for cur_image_type in [self.POSTER, self.BANNER, self.POSTER_THUMB, self.BANNER_THUMB,self.FANART]:
            logger.log(u"Seeing if we still need an image of type " + str(cur_image_type) + ": " + str(
                need_images[cur_image_type]), logger.DEBUG)
            if cur_image_type in need_images and need_images[cur_image_type]:
                self._cache_image_from_indexer(show_obj, cur_image_type)

        logger.log(u"Done cache check")

**Comment by miigotu** [comment]
@cookerrr it is possible the os.path.isFile is failing for you, depending on how the python you have was compiled, and it's version.

This comes about due to the slashes in the file path, and also could be because it is not using the absPath for that check.

Could you try changing this:

```python
                if ek.ek(os.path.isfile, cur_provider.get_poster_path(show_obj)):
                    cur_file_name = os.path.abspath(cur_provider.get_poster_path(show_obj))
```
to:
```python
                cur_file_name = os.path.abspath(r'' + cur_provider.get_poster_path(show_obj))
                if ek.ek(os.path.isfile, cur_file_name):
```

I haven't looked at the code as a whole yet, this is just a guess from the info you provided. From what you have pasted it seems the other image types are also not pulled from meta to cache.

**Comment by cookerrr** [comment]
I tried changing those two lines and the log didn't change and it didn't move the poster and banner. It seems as if the python script halts when it gets to that portion of code, because after the log shows this line "logger.log(u"Checking if we can use the show image from the " + cur_provider.name + " metadata", logger.DEBUG)"  There is no other lines in the log from that script.  When I deleted that section the log showed it pull from tvdb and populate the cache folder.  I'm not sure if this is a problem for anyone else, or if it's just my case.  If so I'm fine with it pulling the artwork straight from thetvdb,  It only has to pull it once when a show is added.  If it helps any I'm running python 2.7.  I just installed sickrage this past week, I'm migrating from sickbeard.  I followed the windows install guide that was linked to in a sticky post on the sickrage.tv forum here http://www.htpcguides.com/install-sickrage-windows-usenet-torrent-tv/

If you have any other ideas, let me know, More than willing to try to make it work the right way!

**Comment by Hiltronix** [comment]
Since I created this thread, I found the issue that resolved the problem, at least it did in my case.  I posted the PR on the SR project here:

https://github.com/SiCKRAGETV/SickRage/pull/2340

And then was told I had to do the PR on the project library that is being used in SR.  So I did, and it was accepted, and can be found here:

https://bitbucket.org/haypo/hachoir/pull-requests/18/hiltronix-when-metadataextractparser/diff

But the fix is still not included in SR.  So once it is, maybe the other people that are having the same or similar issues can test it out, once it's pushed out by the SR devs.  But it wasn't in the last SR update.

**Comment by miigotu** [comment]
@Hiltronix not sure how the images have anything to do with hachoir?

hachoir is only used for getting the resolution of a video, nfo, and torrent files afaik.

@cookerrr if that did in fact fix the error, it would be in the develop branch.Could you try develop and see if it is fixed there?

**Comment by cookerrr** [comment]
The develop branch's images_cache.py file is the same as the master branch.  I could start a new issue about this, but so far I'm wondering if this issue is for some reason isolated to just me.  Removing the portion of that file that just tries to pull the cache locally makes it work perfectly as stated above, and it pulls the images to the cache directory from thetvdb.  I don't know if that's how you would want this to work for everyone.  I would love to fix it so it works like it's meant to by pulling the files from my local metadata, but don't know enough about python to modify the lines to make it work for me.  from my little tinkering It seems as if the script just halts execution around the lines... 
                if ek.ek(os.path.isfile, cur_provider.get_poster_path(show_obj)):
                    cur_file_name = os.path.abspath(cur_provider.get_poster_path(show_obj))
                    cur_file_type = self.which_type(cur_file_name)
I only say this since the log shows none of the print to log lines following that statement.  If you have any other ideas for me to try, I'm more than happy to give it a go.  I did try that modification you posted earlier, but there was no change in what the log file showed. also, thanks for maintaining this wonderful piece of software, It's amazing!

**Comment by miigotu** [comment]
@cookerrr it has nothing to do with image_cache.py being the same. image cache uses hachoir to determine image types, and there was a bug in hachoir effecting this that was fixed.

**Comment by cookerrr** [comment]
ok, just updated to latest master  I noticed  and fiddled with this again.  Another variable in this problem that I didn't think of that may influence the results is that in the config.ini I have create_missing_show_dirs set to 1.  I have kodi delete the show directory when I'm done watching a show.  Now in my couple tests I deleted a banner from sickrages image cache for america's got talent.  The show directory was missing as it was deleted after watching the show.  When I did a force refresh it worked as expected, the log shows it searched for metadata and it wasn't there so it grabbed from thetvdb.  When i tried the same test by deleting a banner from ncis: LA  in which the show directory still exists with the metadata in the show folder, it does the same in log file like it did before, where it stops at "2015-09-06 20:37:55 DEBUG    SHOWQUEUE-REFRESH :: Checking if we can use the show image from the KODI 12+ metadata" and never copies the banner either from the metadata that already exists in the show folder, or grabs the image from thetvdb.

**Comment by cookerrr** [comment]
I just looked at that pull request that hiltronix put in for that hachoir, and noticed he commented two days ago that the problem is still not fixed with that,  so I'll wait this out till they figure out the problem over there before I try this out again.

**Comment by Hiltronix** [comment]
If you're referring to the PR I made a day or two ago at the hachoir project, I'm waiting to see if it will be accepted.  The original PR I made there did fix the problem by eliminating an exception.  But the dev on the project asked me to add the error logger function to the exception handler I added.  And it turned out to add the exception issue right back into the situation.  So after more research it turns out that the exception is being caused by the error logging function.  The line "sys.stderr.flush()" is causing an exception that doesn't allow "extractMetadata" to exit properly, and generates the "[Errno 9] Bad file descriptor" exception error.

**Comment by miigotu** [comment]
@Hiltronix check the type of the error variable, stderr issues are usually due to encoding. I think we are passing it a unicode filepath/name, and flush would be trying to encode it to ascii I think.

**Comment by Hiltronix** [comment]
The type of the error variable is "str" not "unicode".

The actual error messages from a "Re-Scan" on the 2 shows with missing artwork that I use to test are as follows:

The Grinder - Rescan

prefix: [warn]
text: [/exif/content] [Autofix] Fix parser error: stop parser, found unparsed segment: start 1392, length 16, found unparsed segment: start 1744, length 16


2 Broke Girls - Rescan

prefix: [warn]
text: Unable to convert creation_date=u'2013-11-09T12:19:43+02:00' (unicode) to datetime or date

prefix: [warn]
text: [/exif/content] [Autofix] Fix parser error: stop parser, found unparsed segment: start 1776, length 48

prefix: [warn]
text: Unable to convert creation_date=u'2013-11-09T12:19:43+02:00' (unicode) to datetime or date

prefix: [warn]
text: [/exif/content] [Autofix] Fix parser error: stop parser, found unparsed segment: start 1776, length 48

**Comment by fernandog** [comment]
@miigotu confirmed issue?

**Comment by parneye** [comment]
I believe I have a rather simple temporary workaround to the problem.

In MS Win7, it appears the issue only occurs if you launch SickRage with "pythonw.exe" (I tried version 2.7.5 and 2.7.10)
If you launch with "python.exe" instead it adds the pictures to cache correctly and works well.

I did some further debugging, not sure if this is handy to the developers.. I've attached logs for where it stops when running with pythonw.exe, and also the behaviour with python.exe for reference. 
<code>
## RUNNING WITH PYTHONW.exe
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Metadata provider KODI 12+ creating show metadata for The 100
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Writing show nfo file to E:\fakey\TV Shows\The 100\tvshow.nfo
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: 268592: Writing NFOs for all episodes
2015-10-10 14:43:18 INFO     SHOWQUEUE-ADD :: 268592: Updating NFOs for show with new indexer info
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Checking if E:\fakey\TV Shows\The 100\tvshow.nfo exists: True
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Metadata provider KODI 12+ updating show indexer info metadata file for The 100
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Checking & filling cache for show The 100

## RUNNING WITH PYTHON.exe
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Metadata provider KODI 12+ creating show metadata for The 100
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Writing show nfo file to E:\fakey\TV Shows\The 100\tvshow.nfo
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: 268592: Writing NFOs for all episodes
2015-10-10 14:43:18 INFO     SHOWQUEUE-ADD :: 268592: Updating NFOs for show with new indexer info
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Checking if E:\fakey\TV Shows\The 100\tvshow.nfo exists: True
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Metadata provider KODI 12+ updating show indexer info metadata file for The 100
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Checking & filling cache for show The 100
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Checking if we need any cache images for show 268592
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Checking if file c:\myfakefolder\SickRangeRunning\SRv1\cache\images\268592.poster.**g exists
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Checking if file c:\myfakefolder\SickRangeRunning\SRv1\cache\images\268592.banner.**g exists
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Checking if file c:\myfakefolder\SickRangeRunning\SRv1\cache\images\thumbnails\268592.poster.**g exists
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Checking if file c:\myfakefolder\SickRangeRunning\SRv1\cache\images\thumbnails\268592.banner.**g exists
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Checking if file c:\myfakefolder\SickRangeRunning\SRv1\cache\images\268592.fanart.**g exists
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Checking if we can use the show image from the KODI 12+ metadata
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Checking if image E:\fakey\TV Shows\The 100\poster.**g (type 2 needs metadata: True
2015-10-10 14:43:18 DEBUG    SHOWQUEUE-ADD :: Found an image in the show dir that doesn't exist in the cache, caching it: E:\fakey\TV Shows\The 100\poster.**g, type 2
2015-10-10 14:43:18 INFO     SHOWQUEUE-ADD :: Copying from E:\fakey\TV Shows\The 100\poster.**g to c:\myfakefolder\SickRangeRunning\SRv1\cache\images\268592.poster.**g
</code>

**Comment by OmgImAlexis** [comment]
@parneye could you add ```` to the top and bottom of your logs like this.

![screen shot 2015-10-10 at 6 51 10 pm](https://cloud.githubusercontent.com/assets/6525926/10410132/e7603fa2-6f7f-11e5-8a4a-5fc216157e22.png)

**Comment by OmgImAlexis** [comment]
Also there's a reason pythonw.exe is used. https://stackoverflow.com/questions/9705982/pythonw-exe-or-python-exe

**Comment by parneye** [comment]
thanks, sorry about the incorrect formatting. Regarding python/pythonw I prefer a minimised console window if it comes with working posters/banners than no console window but also no posters.. unless I missed a more important difference in that link.. Not completely sure, but I think one the mentioned differences is why it does side step the issue. With pythonw (Unhandled exceptions cause the script to abort silently.)..

**Comment by miigotu** [comment]
Do not use pythonw.exe. open a prompt and run python.exe in it if you need console output.

**Comment by parneye** [comment]
hi miigotu. I think you may have meant to say "if you need images displayed on the show list" as the issue is not the lack of console output, rather that images don't copy to the cache folder when sickrage is executed with pythonw.exe
Are you able to re-open this issue, or maybe rename it to reference pythonw.exe? I'm not sure of the requirements to be allowed to edit code for the project, but I would like to work on an elegant solution as pythonw is the default/nicest way to run python in windows. I'll look in to either a way to minimise the console window to the system-tray or modifications to the code where it is dropping out.

**Comment by miigotu** [comment]
`pythonw is the default/nicest way to run python in windows`

Totally incorrect.

**Comment by parneye** [comment]
okay, I stand corrected. what I meant to say was that it's the default used when launching a python file in windows if you have installed python for windows with the default settings.. at least I think it is...
I'm pretty new to python, and I have a lot more reading to do -- I'm a c# developer and I became interested in python after I started using sickrage.
do you know where the issue resides with this? I'm curious if it's something inherent in pythonw or something else? personally I only experienced the show list image display issue when tvrage was removed.

**Comment by miigotu** [comment]
Just dont use pythonw. SickRage is a daemon application. pythonw is meant for cli alpplications.

Likely the problem is that you are not setting the encoding when running it, so any unicode that pops into the console will crash pythonw.

This __might__ work, but is unsupported.
`chcp 65001 && pythonw.exe SickBeard.py`

**Comment by parneye** [comment]
Okay thanks, I'll have a look at it. When you say CLI do you mean something different than usual? As python.exe lets you interact within the terminal/command window whereas pythonw launches the application as a daemon/service.
I did a bit more investigation and resolved with replacing the use of hachoir with pillow:
the issue arises when the ```which_type``` method in image_cache.py calls the hachoir_metadata extractMetadata function. Installing Pillow ```pip install Pillow``` and modifying the ```which_type``` method in ```image_cache.py``` to use it resolves the issue. Below is the code to copy in after commenting out the relevant section.
###### Section of code to comment out
```python
# use hachoir to parse the image for us
img_parser = createParser(path)
img_metadata = extractMetadata(img_parser)
if not img_metadata:
    logger.log(u"Unable to get metadata from " + str(path) + ", not using your existing image", logger.DEBUG)
    return None
img_ratio = float(img_metadata.get('width')) / float(img_metadata.get('height'))
img_parser.stream._input.close()
```
###### Section of code to insert
```
from PIL import Image
im = Image.open(path)
width, height = im.size
im.close()
img_ratio = float(width) / float(height)
````
This works well, I've tested adding about 20 shows. I haven't made any exception handling code yet.

**Comment by miigotu** [comment]
Open SickRage/lib/hachoir_core/log.py and change `self.use_print = True` to `self.use_print = False`

Let me know if that fixes the problem. I am guessing it is the `[warn] skip parser: false` printed by hachoir on parse error.

**Comment by cookerrr** [comment]
so hearing changing to python.exe fixed this problem, I did change mine from pythonw to python.exe and that fixed the problem for me as well.  If I had known this before I would have been happy with that solution.  I just installed sickrage as a system service so the console window never shows.  While I don't understand why pythonw is so different as I know nothing about python, I'm just happy this is working.  If pythonw just won't work for whatever reason, I'd say just to document it somewhere to say to not use pythonw as it could cause problems, and only use python.exe to run sickrage.
